rust/src/dataset/fragment.rs

```diff
@@ -199,6 +202,58 @@ impl FileFragment {
 
         Ok(Updater::new(self.clone(), reader))
     }
+
+    /// Delete rows from the fragment.
+    ///
+    /// If all rows are deleted, returns `Ok(None)`. Otherwise, returns a new
+    /// fragment with the updated deletion vector. This must be persisted to
+    /// the manifest.
+    pub(crate) async fn delete(mut self, predicate: &str) -> Result<Option<Self>> {
+        // Load existing deletion vector
+        let mut deletion_vector = read_deletion_file(&self.metadata, self.dataset.object_store())
+            .await?
+            .unwrap_or_default();
+
+        // scan with predicate and row ids
+        let mut scanner = self.scan();
+        scanner
+            .with_row_id()
+            .filter(predicate)?
+            .project::<&str>(&[])?;
+
+        // As we get row ids, add them into our deletion vector
+        scanner
+            .try_into_stream()
+            .await?
+            .try_for_each(|batch| {
+                let array = batch["_rowid"].clone();
+                let int_array: &UInt64Array = as_primitive_array(array.as_ref());
+
+                // _row_id is global, not within fragment level. The high bits
+                // are the fragment_id, the low bits are the row_id within the
+                // fragment.
+                let local_row_ids = int_array.iter().map(|v| v.unwrap() as u32);
+
+                deletion_vector.extend(local_row_ids);
+                futures::future::ready(Ok(()))
+            })
+            .await?;
+
+        let num_rows = self.count_rows().await? as u32;
+        if deletion_vector.contains_range(0..num_rows) {
```

Comment:
nit: we could do something like `deletion_vector.len() == num_rows`, then iterating over all the items

Reply:
That's a good idea. This is potentially slow in the common case where not all rows are deleted.